papis/commands/update.py

```diff
@@ -86,100 +369,163 @@ def run(document: papis.document.Document,
-                "An importer set the 'ref' key. This is not allowed and will be "
-                "automatically removed. Check importers: '%s'",
-                "', '".join(importer.name for importer in matching_importers))
-
-            del imported.data["ref"]
-
-        ctx.data.update(imported.data)
-
-        run(document, data=ctx.data, git=git)
+        if to_set:
+            run_set(ctx.data, to_set)
+
+        if to_append and success:
+            success = run_append(ctx.data, to_append, KEY_TYPES)
+
+        if to_remove and success:
+            success = run_remove(ctx.data, to_remove)
+
+        if to_drop and success:
+            run_drop(ctx.data, to_drop)
+
+        if to_rename:
+            success = run_rename(ctx.data, to_rename)
+
+        if success:
+            logger.info(
+                "Updating {c.Back.WHITE}{c.Fore.BLACK}%s{c.Style.RESET_ALL}.",
+                papis.document.describe(document),
+            )
+
+            # NOTE: use 'papis addto' to add files, so this only adds data
+            # by setting 'only_data' to True always
+            matching_importers = papis.utils.get_matching_importer_by_name(
+                from_importer, only_data=True
+            )
+
+            if not from_importer and auto:
+                for importer_cls in papis.importer.get_importers():
+                    try:
+                        importer = importer_cls.match_data(document)
+                        if importer:
+                            try:
+                                importer.fetch_data()
+                            except NotImplementedError:
+                                importer.fetch()
+                    except NotImplementedError:
+                        continue
+                    except Exception as exc:
+                        logger.exception("Failed to match document data.", exc_info=exc)
+                    else:
+                        if importer and importer.ctx:
+                            matching_importers.append(importer)
+
+            imported = papis.utils.collect_importer_data(
+                matching_importers, batch=batch, only_data=True
+            )
+            if "ref" in imported.data:
+                logger.debug(
+                    "An importer set the 'ref' key. This is not allowed and will be "
+                    "automatically removed. Check importers: '%s'",
+                    "', '".join(importer.name for importer in matching_importers),
+                )
+
+                del imported.data["ref"]
+
+            ctx.data.update(imported.data)
+
+            # Delete keys that don't have any values
+            {k: v for k, v in ctx.data.items() if v}
+
+            processed_documents.append((document, ctx.data))
+
+        if not success:
+            processed_documents.clear()
+            logger.error(
+                "Aborting operation. No documents have been changed.",
+            )
+            break
```

Comment:
I'm not sure I understand this success flag? Why is it aborting the whole thing if one document failed?

Reply:
Just noticed that we already discussed this a bit in https://github.com/papis/papis/pull/681#discussion_r1336995439 and the same logic would apply here. Fair enough, I don't think I have any better suggestions here, but I may complain when I hit a use case where it would be nice to just steamroll through any errors :grin:

Reply:
Oh, maybe steamroll through errors if `--batch` is set? That sounds like a "just do it!" sort of flag..

Reply:
Can you explain me your usecase? I was thinking that if there's some error then clearly you did something wrong. I guess maybe your library is inconsistent and you just want to get something done rather than fix it? 

I can live with the suggestions re steamrolling through errors when `--batch` is set though :laughing:

Reply:
What I had in mind was some more general clean up of some sort, e.g.
```
papis update --append tags --all <QUERY>
```
and one of the documents has `tags` as a string instead of as a list, so the whole thing would fail. But you're right, this is probably best solved by going in an fixing that document.